docs/function-inlining.md

```diff
@@ -0,0 +1,40 @@
+# No support for user inlining (yet)
+
+## Summary
+<!-- https://luau-lang.org/performance#function-inlining-and-loop-unrolling -->
+Luau's bytecode compiler (or VM?) is smart about inlining functions. We would rather focus on improving this automation than risk the chance of its misuse.
+
+## Motivation
```

Comment:
I would like to see more cases where `@inline` doesn't apply:

1. Exported without a `table.freeze`.
2. "Object oriented" code.
3. Metamethod.

So 1 would look something like:

```luau
local mod = {}

function mod.sum(lst: {number})
  return mod.sumAcc(0, 1, lst)
end

function mod.sumAcc(acc: number, i: number, lst: {number})
  return if i > #lst then acc else mod.sumAcc(acc + lst[i], i + 1, lst)
end
```

Reply:
Oh, for pt 1, I want to mention that even if you added `table.freeze`, there's no guarantee that the property pointing to the function implementation did not get transmuted to point to a different function _before_ the table got frozen.

There might be a simpler example, but first one that comes to mind:

```luau
-- A.luau
local mod = {}

function mod.print_foo(t)
  print(t.foo)
end

return mod

-- B.luau
local A = require("./A")

local mod = {}

function mod.foo()
  -- `mod.foo` could theoretically be called without our knowledge
  -- which in turn is passed into `A.foo`, which we can't see the implementation of
  -- and so we must assume it mutates `mod`, making `mod.foo` not inlineable.
  A.foo(mod)
end

return table.freeze(mod)
```